### PR TITLE
Initial Commit, used stack new

### PR DESCRIPTION
**Pull request by johnny555**
Largely a cut and paste from the notebook https://github.com/johnny555/ToolExamples/blob/master/Kaggle/Titanic/H-SMO.ipynb

Interested to know:
- What the right way to split the module is. (I.e. should I keep data declarations in their own file?)
- Repa and ComputeS, is there a better way?
- Any code smells at this stage?

Note, the code only typechecks... still early days....

### CODE REVIEW

**Comment by johnny555** [comment]
@NickSeagull @Drezil @SamuelSchlesinger 
sorry guys, I had meant to add you in the pull request!

**Comment by NickSeagull** [comment]
Notifying @DataHaskell/everyone so people can comment more 😄

**Comment by ocramz** [comment]
@johnny555 I find it a bit hard to read.. possibly because the lens and repa notation are interleaved with the equations. But what matters now is that it works! : D

File: src/Numeric/Algorithms/HSVO/Detail.hs [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+              -> Value
+lowerAlpha params sv1 sv2 =
+    let
+        y1 = sv1^.trueLabel
+        y2 = sv2^.trueLabel
+        a1 = scalarToDbl $ sv1^.supvec.alpha
+        a2 = scalarToDbl $ sv2^.supvec.alpha
+        c = scalarToDbl $ params^.margin
+    in
+        if not(y1 == y2)
```
Comment (axman6) [inline src/Numeric/Algorithms/HSVO/Detail.hs]: `y1 /= y2`

File: src/Numeric/Algorithms/HSVO/SVM.hs [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    vecs <- get
+    params <- ask
+    pure $ \x -> collateMaybes x [takeStepDetail params vec x  | vec <-  (_vectorList vecs ) ] -- list comprehension to evaluate over all maybes...
+
+helper1 :: Maybe a -> [a] -> [a]
+helper1 (Just a) b = a : b
+helper1 Nothing b = b
+
+maybeToLists :: [Maybe a] -> [a]
+maybeToLists  =
```
Comment (axman6) [inline src/Numeric/Algorithms/HSVO/SVM.hs]: `catMaybes`

File: src/Numeric/Algorithms/HSVO/SVM.hs [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+
+
+collateMaybes :: TrainingSupportVector -> [Maybe (TrainingSupportVector, TrainingSupportVector)] -> Maybe TrainingSupportVector
+collateMaybes tsv1 tsvList =
+  let
+     validResults = WorkingState {_vectorList = snd <$> maybeToLists tsvList}
+     alpha_data = foldrOf (vectorList . traverse . supvec . alpha) (\a (sum, count) -> ( (scalarToDbl a) + sum, count+1)) (0, 0) validResults
+     sum =  fst alpha_data
+     len =  snd alpha_data
```
Comment (axman6) [inline src/Numeric/Algorithms/HSVO/SVM.hs]: `(sum,len) = alpha_data`

Also take a look at the `foldl` package.

File: src/Lib.hs [hunk 4]
```diff
@@ -0,0 +1,6 @@
+module Lib
```
Comment (johnny555) [inline src/Lib.hs]: ok

File: src/Numeric/Algorithms/HSVO/Detail.hs [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+
+
+import Data.Array.Repa.Repr.Vector
+import qualified Data.Vector as V
+
+import Data.Array.Repa.Repr.Vector                   as RV
+import Data.Array.Repa.Algorithms.Matrix
+
+type Value = Double
+
```
Comment (johnny555) [inline src/Numeric/Algorithms/HSVO/Detail.hs]: My reasoning here is that I'm not sure about trade offs wrt performance for different representations, so I thought I would write it such that the decision of what underlying value to use could be decided later. But you are right, Double is definately more understandable. 
Do you think I should change it back to double?

File: src/Numeric/Algorithms/HSVO/Detail.hs [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+type BaseVector = Array U DIM1 Value
+type BaseScalar = Array U Z Value
```
Comment (johnny555) [inline src/Numeric/Algorithms/HSVO/Detail.hs]: Yeah, this has turned out to be abit of a headache as I'm building out the SVM side of things. However if I want to be able to use the Repa operators, then the variables need to be inside of repa objects I think. 
I am starting to wonder if Repa was a good design choice...

File: src/Numeric/Algorithms/HSVO/SVM.hs (hunk 3 above)
Comment (johnny555) [inline src/Numeric/Algorithms/HSVO/SVM.hs]: cool, thanks.
Do you think I should replace the lens foldOf with something from the foldl pacakage? or is there something I'm doing here that could benefit from combinators over folds?

File: src/Numeric/Algorithms/HSVO/Detail.hs (hunk 1 above)
Comment (johnny555) [inline src/Numeric/Algorithms/HSVO/Detail.hs]: ok

File: src/Numeric/Algorithms/HSVO/Detail.hs [hunk 7]
```diff
@@ -0,0 +1,489 @@
...
+                   -> a                -- ^ L
+                   -> a
+alphaNewClipped a h l
+    | a >= h = h
+    | a <= l = l
+    | otherwise = a
+
+
+scalarToDbl :: BaseScalar -> Value
+scalarToDbl s = s ! ( Z )
```
Comment (johnny555) [inline src/Numeric/Algorithms/HSVO/Detail.hs]: ok

File: src/Numeric/Algorithms/HSVO/Detail.hs [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+getLabelValue :: PredictedLabel -> Value
+getLabelValue (PredictClass1 v) = v
+getLabelValue (PredictClass2 v) = v
+
+predictToTrue :: PredictedLabel -> ClassLabel
+predictToTrue (PredictClass1 _) = Class1
+predictToTrue (PredictClass2 _) = Class2
+
+wrapScalar :: Value -> BaseScalar
+wrapScalar s = fromListUnboxed Z ([s] :: [Value])
```
Comment (johnny555) [inline src/Numeric/Algorithms/HSVO/Detail.hs]: ok

**Comment by johnny555** [comment]
Thanks @axman6 I've changed a bunch of stuff. Yes I agree there is still work to do in putting the overall algorithm together. 
Yeah, I'm not sure how everything is going to land performance wise, but I am finding lenses are making life nice, and much of my design is around things I can do with them. Although, if lenses turn out to be an optimisation killer then I might have to do a very large refactor to remove them.

File: LICENSE [hunk 9]
```diff
@@ -1,21 +1,30 @@
-MIT License
+Copyright Author name here (c) 2016
```
Comment (stites) [inline LICENSE]: Seems like the license got overridden

**Comment by ocramz** [comment]
Hi @johnny555 , I merged just to make your work more visible, I hope you don't mind! Thanks for pushing this forward!

**Comment by johnny555** [comment]
hey thanks @ocramz ! I've been working away at this for alittle while, but I'm going to have to put it on hold for a bit and thought I would upload where I was at. It compiles and everything is linked up. Not sure if it actually works though, further testing required. 
Cheers!

**Pull request merged**